### ISSUE TITLE
Error when mixing sync and async tasks

### ISSUE THREAD

**Issue by macro128** [PrefectHQ/prefect#6318]
### First check

- [x] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Prefect documentation for this issue.
- [X] I checked that this issue is related to Prefect and not one of its dependencies.

### Bug summary

Hi! There's a bug when using sync tasks and async tasks in the same async flow, an exception related to the event loop is raised when a sync tasks is submitted (using `.submit()`).
There's also a workaround using only async tasks (even when only sync code is executed) but it would be nice being able to mix both in the same flow.

### Reproduction

```python
import asyncio

from prefect import task, get_run_logger, flow


@task
async def task1():
    await asyncio.sleep(1)


@task
def task2():
    logger = get_run_logger()
    logger.info("hi")


@flow
async def test_flow():
    await task1.submit()
    task2.submit() # raises here
```

### Error

```python
Encountered exception during execution:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/prefect/engine.py", line 550, in orchestrate_flow_run
    result = await flow_call()
  File "/app/src/streamcast/agent/flows/test.py", line 22, in test_flow
    task2.submit()
  File "/opt/conda/lib/python3.10/site-packages/prefect/tasks.py", line 491, in submit
    return enter_task_run_engine(
  File "/opt/conda/lib/python3.10/site-packages/prefect/engine.py", line 691, in enter_task_run_engine
    return flow_run_context.sync_portal.call(begin_run)
  File "/opt/conda/lib/python3.10/site-packages/anyio/from_thread.py", line 283, in call
    return cast(T_Retval, self.start_task_soon(func, *args).result())
  File "/opt/conda/lib/python3.10/concurrent/futures/_base.py", line 439, in result
    return self.__get_result()
  File "/opt/conda/lib/python3.10/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
  File "/opt/conda/lib/python3.10/site-packages/anyio/from_thread.py", line 219, in _call_func
    retval = await retval
  File "/opt/conda/lib/python3.10/site-packages/prefect/engine.py", line 783, in create_task_run_then_submit
    task_run = await create_task_run(
  File "/opt/conda/lib/python3.10/site-packages/prefect/engine.py", line 823, in create_task_run
    task_run = await flow_run_context.client.create_task_run(
  File "/opt/conda/lib/python3.10/site-packages/prefect/client.py", line 1635, in create_task_run
    response = await self._client.post(
  File "/opt/conda/lib/python3.10/site-packages/httpx/_client.py", line 1842, in post
    return await self.request(
  File "/opt/conda/lib/python3.10/site-packages/httpx/_client.py", line 1527, in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
  File "/opt/conda/lib/python3.10/site-packages/prefect/client.py", line 258, in send
    await super().send(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/httpx/_client.py", line 1614, in send
    response = await self._send_handling_auth(
  File "/opt/conda/lib/python3.10/site-packages/httpx/_client.py", line 1642, in _send_handling_auth
    response = await self._send_handling_redirects(
  File "/opt/conda/lib/python3.10/site-packages/httpx/_client.py", line 1679, in _send_handling_redirects
    response = await self._send_single_request(request)
  File "/opt/conda/lib/python3.10/site-packages/httpx/_client.py", line 1716, in _send_single_request
    response = await transport.handle_async_request(request)
  File "/opt/conda/lib/python3.10/site-packages/httpx/_transports/default.py", line 353, in handle_async_request
    resp = await self._pool.handle_async_request(req)
  File "/opt/conda/lib/python3.10/site-packages/httpcore/_async/connection_pool.py", line 253, in handle_async_request
    raise exc
  File "/opt/conda/lib/python3.10/site-packages/httpcore/_async/connection_pool.py", line 237, in handle_async_request
    response = await connection.handle_async_request(request)
  File "/opt/conda/lib/python3.10/site-packages/httpcore/_async/connection.py", line 90, in handle_async_request
    return await self._connection.handle_async_request(request)
  File "/opt/conda/lib/python3.10/site-packages/httpcore/_async/http11.py", line 105, in handle_async_request
    raise exc
  File "/opt/conda/lib/python3.10/site-packages/httpcore/_async/http11.py", line 84, in handle_async_request
    ) = await self._receive_response_headers(**kwargs)
  File "/opt/conda/lib/python3.10/site-packages/httpcore/_async/http11.py", line 148, in _receive_response_headers
    event = await self._receive_event(timeout=timeout)
  File "/opt/conda/lib/python3.10/site-packages/httpcore/_async/http11.py", line 177, in _receive_event
    data = await self._network_stream.read(
  File "/opt/conda/lib/python3.10/site-packages/httpcore/backends/asyncio.py", line 33, in read
    return await self._stream.receive(max_bytes=max_bytes)
  File "/opt/conda/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 1265, in receive
    await self._protocol.read_event.wait()
  File "/opt/conda/lib/python3.10/asyncio/locks.py", line 211, in wait
    fut = self._get_loop().create_future()
  File "/opt/conda/lib/python3.10/asyncio/mixins.py", line 30, in _get_loop
    raise RuntimeError(f'{self!r} is bound to a different event loop')
RuntimeError: <asyncio.locks.Event object at 0x7fd237ee4310 [unset]> is bound to a different event loop
```

### Versions

```
Version:             2.0.3
API version:         0.8.0
Python version:      3.10.5
Git commit:          2f1cf4ac
Built:               Fri, Aug 5, 2022 3:57 PM
OS/Arch:             linux/x86_64
Profile:             default
Server type:         hosted
```

**Comment by sti0** [comment]
@madkinsz is there any timeline when this problem gets fixed?

**Comment by madkinsz** [comment]
I can't provide a timeline but this is my primary project. I'm exploring some holistic solutions to similar issues. I expect pull requests will start coming in the next few weeks.

**Comment by zhen0** [comment]
Adding some notes on here as I'm running through and checking issues. 

On Prefect 2.8.6 I get the following when running the MRE flow locally:
```
 RuntimeWarning: coroutine 'create_then_begin_flow_run' was never awaited
  test_flow()
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```
And on Prefect 2.8.7:
```
 RuntimeWarning: coroutine 'from_async.wait_for_call_in_loop_thread' was never awaited
  test_flow()
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

**Comment by zhen0** [comment]
If I use a deployment and agent, the MRE flow runs to completion.

**Comment by madkinsz** [comment]
Addressed by #8702 — MRE requires `asyncio.run(test_flow())`
